rig-core/src/completion/message.rs

```diff
@@ -0,0 +1,488 @@
+use std::{convert::Infallible, str::FromStr};
+
+use crate::OneOrMany;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use super::CompletionError;
+
+// ================================================================
+// Request models
+// ================================================================
+
+/// A message represents a run of input (user) and output (assistant).
+/// Each message type (based on it's `role`) can contain a atleast one bit of content such as text,
+///  images, audio, documents, or tool related information. While each message type can contain
+///  multiple content, most often, you'll only see one content type per message
+///  (an image w/ a description, etc).
+///
+/// Each provider is responsible with converting the generic message into it's provider specific
+///  type using `From` or `TryFrom` traits. Since not every provider supports every feature, the
+///  conversion can be lossy (providing an image might be discarded for a non-image supporting
+///  provider) though the message being converted back and forth should always be the same.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+#[serde(tag = "role", rename_all = "lowercase")]
+pub enum Message {
+    /// User message containing one or more content types defined by `UserContent`.
+    User { content: OneOrMany<UserContent> },
+
+    /// Assistant message containing one or more content types defined by `AssistantContent`.
+    Assistant {
+        content: OneOrMany<AssistantContent>,
+    },
+}
+
+/// Describes the content of a message, which can be text, a tool result, an image, audio, or
+///  a document. Dependent on provider supporting the content type. Multimedia content is generally
+///  base64 (defined by it's format) encoded but additionally supports urls (for some providers).
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum UserContent {
+    Text(Text),
+    ToolResult(ToolResult),
```

Comment:
this type should be part of `Message` enum as it's a tool generated message (not user generated)

Reply:
earlier convo